treescript/src/treescript/merges.py

```diff
@@ -44,22 +59,35 @@ def touch_clobber_file(repo_path):
 async def apply_rebranding(config, repo_path, merge_config):
     """Apply changes to repo required for merge/rebranding."""
     log.info("Rebranding %s to %s", merge_config.get("from_branch"), merge_config.get("to_branch"))
-    if merge_config.get("version_files"):
-        current_version = get_version("browser/config/version.txt", repo_path)
-        next_version = f"{current_version.major_number}.{current_version.minor_number}"
 
+    version = get_version("browser/config/version.txt", repo_path)
+    current_major_version = version.major_number
+    if merge_config.get("incr_major_version", False):
+        version = version.bump("major_number")
+
+    if merge_config.get("version_files"):
+        next_version = f"{version.major_number}.{version.minor_number}"
         await do_bump_version(config, repo_path, merge_config["version_files"], next_version)
-    if merge_config.get("version_files_suffix"):
-        current_version = get_version("browser/config/version.txt", repo_path)
-        current_version = attr.evolve(current_version, is_esr=False, beta_number=None, is_nightly=False)
-        next_version = f"{current_version}{merge_config.get('version_suffix')}"
 
+    if merge_config.get("version_files_suffix"):
+        version = attr.evolve(version, is_esr=False, beta_number=None, is_nightly=False)
+        next_version = f"{version}{merge_config.get('version_suffix')}"
         await do_bump_version(config, repo_path, merge_config["version_files_suffix"], next_version)
 
     for f in merge_config.get("copy_files", list()):
         shutil.copyfile(os.path.join(repo_path, f[0]), os.path.join(repo_path, f[1]))
 
+    format_options = {
+        "current_major_version": current_major_version,
+        "next_major_version": version.major_number,
+        "current_weave_version": current_major_version + 2,
+        "next_weave_version": current_major_version + 3,  # current_weave_version + 1
+    }
+
+    fmt = BashFormatter()
```

Comment:
Question: What does this custom formatter achieve? I'm trying to understand what it does compared to https://docs.python.org/3/library/shlex.html#shlex.quote

Reply:
In order to deal with formatting like `{next_major_version}` we have to to string formatting, but some of the file placements contain bash variables, such  as this one: `"MOZ_REQUIRE_SIGNING=${MOZ_REQUIRE_SIGNING-0}"` 
The default `.format()` will throw a `KeyError` there. `shlex` is about making things shell-safe

So what we need is a formatter that will fill in variables it has been told about, such as `next_major_version` but not error if it hasn't been told about `MOZ_REQUIRE_SIGNING` because we want that passed through literally.

My first thought was to simply double `{{}}` the bash variables but that would mean that anyone looking for matching strings with grep would miss these - I can imagine a situation where we rename `MOZ_REQUIRE_SIGNING` and someone does `grep -r '${MOZ_REQUIRE_SIGNING}'` to find other references to it, and the `{{` `}}` would mean the file replacement entry is missed. 

Since `.format` can't be told to ignore missing keys, we need a formatter that can be told what to do, and this does the job: If it's not passed a value for the `{key}` then it returns the key, letting the value be unchanged.

Phew that was long, but I hope it helps

Reply:
It helps a lot! Thank you very much for this thorough answer! 

Let's add a code comment to include this piece of knowledge alongside the class. This way, future readers will know right away 😃